### PR TITLE
Improve movement

### PR DESCRIPTION
**Pull request by Feesh7**
Adds:
- Variable jump height based on jump button press length
- Optional boost in speed when jumping while running/wall running
- Camera chases player now instead of being fixed
- Z targeting
- Button class

### CODE REVIEW

File: Assets/Scripts/PlayerController.cs [hunk 1]
```diff
@@ -940,10 +929,69 @@ public Vector3 GetLastWallNormal()
...
+        // Add additional gravity when going down (optional)
+        if (current_velocity.y < 0)
+        {
+            GravityMult += DownGravityAdd;
+        }
+
+        // Handle jumping and falling
+        if (willJump)
+        {
+            DoJump();
```
Comment (nayrpg) [inline Assets/Scripts/PlayerController.cs]: Sloppy!

File: Assets/Scripts/PlayerController.cs [hunk 2]
```diff
@@ -974,6 +1022,9 @@ private void DoJump()
             {
                 //Debug.Log("Wall Run Jump");
                 current_velocity += PreviousWallNormal * WallRunJumpSpeed * JumpMeterComputed;
+                float pathvel = Vector3.Dot(current_velocity, transform.forward);
+                float newspeed = Mathf.Clamp(pathvel + (WallRunJumpBoostAdd * JumpMeterComputed), 0f, WallRunJumpBoostSpeed);
```
Comment (nayrpg) [inline Assets/Scripts/PlayerController.cs]: What do Math Clamps have to do with jumping???

File: Assets/Scripts/PlayerController.cs [hunk 3]
```diff
@@ -974,6 +1022,9 @@ private void DoJump()
             {
                 //Debug.Log("Wall Run Jump");
                 current_velocity += PreviousWallNormal * WallRunJumpSpeed * JumpMeterComputed;
```
Comment (nayrpg) [inline Assets/Scripts/PlayerController.cs]: Now we have Jump Meters? What do I have to charge my feet???

File: Assets/Scripts/PlayerController.cs [hunk 4]
```diff
@@ -23,20 +23,26 @@ public class PlayerController : MonoBehaviour {
...
+    public float JumpBoostAdd;
     public Vector3 StartPos;
     [Header("Movement toggles")]
     public bool wallRunEnabled;
     public bool wallJumpEnabled;
     public bool wallClimbEnabled;
     public bool conserveUpwardMomentum;
+    public bool ShortHopEnabled;
+    public bool JumpBoostEnabled;
     [HideInInspector]
```
Comment (nayrpg) [inline Assets/Scripts/PlayerController.cs]: You better hide it. No one wants to see all this garbage code

File: Assets/Scripts/InputManager.cs [hunk 5]
```diff
@@ -1,8 +1,52 @@
...
+        _active = true;
+        _button_names = button_names;
+    }
+
+    public bool down()
+    {
+        if (_active)
+        {
+            _active = false;
+            return _button_names.Any(name => Input.GetButtonDown(name));
```
Comment (nayrpg) [inline Assets/Scripts/InputManager.cs]: Ohhhhhh so now any name is better than the names for our buttons, huh?!?

**Comment by nayrpg** [comment]
I guess bad code is better than no code...

**Comment by Feesh7** [comment]
¯\_(ツ)_/¯ I'll use more gotos next time

**Pull request merged**